File: lib/agents/adjacent.ts
import { searchWeb, searchNews } from '../tools/serpapi';
import { getTechSentiment } from '../tools/hn-algolia';
import { searchReddit } from '../tools/reddit';
import { planQueries } from '../tools/query-planner';
import { filterRelevant, requirementsFromContext } from '../tools/relevance';
import { extractCandidates, verifyCandidates } from '../tools/candidate-discovery';
import { generateHuggingFaceJson } from './gemini';
import type {
  AgentConfig,
  AgentContext,
  AgentOutput,
  AdjacentOutput,
  AdjacentThreat,
  AgentSource,
  ConfidenceLevel,
  EvidenceCandidate,
} from './types';
import { scoreToLevel } from './types';
import { computeSignalQualityPenalty, extractToolResults } from '../tools/fallback';
import { runResearchLoop } from './research-loop';
import { applyInsufficientGate, evidencePromptRules } from './evidence-gate';
import { localeFromGeography } from './search-locale';
import {
  resolveEntityProbes,
  resolveGapQueries,
  resolveSearchQueries,
} from './plan-queries';
import type { SearchResult, ToolResult } from '../tools/types';
import type { discoverAndScrape } from '../tools/discover-and-scrape';

async function run(ctx: AgentContext): Promise<AgentOutput> {
  const { query, product, competitor, priorContext, geography, category: cat, namedEntities, requiredTerms } = ctx;

  const locale = localeFromGeography(geography);
  const requirements = requirementsFromContext(ctx);

  const queryBundle = planQueries({
    product,
    competitor,
    domain: 'adjacent',
    query,
    category: cat,
    geography,
    namedEntities,
    requiredTerms,
  });

  const planned = resolveSearchQueries(ctx, [queryBundle.broad, queryBundle.targeted], 2);
  const primaryQuery = planned[0] ?? queryBundle.broad;
  const secondaryQuery = planned[1] ?? queryBundle.hypothesis;
  const entityProbes = resolveEntityProbes(ctx, queryBundle.entityProbes, 2);
  const gapQueries = resolveGapQueries(ctx, 3);

  const categoryLabel = cat || (competitor ? `${product} vs ${competitor}` : product);
  const searchedFor = [
    primaryQuery,
    secondaryQuery,
    queryBundle.hypothesis,
    ...entityProbes,
    ...gapQueries,
  ];
  let candidates: EvidenceCandidate[] = [];

  const loop = await runResearchLoop({
    domain: 'adjacent',
    requirements,
    budgetMs: 15_000,
    round1: () =>
      Promise.allSettled([
        searchWeb(primaryQuery, locale),
        searchWeb(secondaryQuery, locale),
        searchWeb(queryBundle.hypothesis, locale),
        searchNews(primaryQuery, locale),
        getTechSentiment(`${product} disruption threat`),
        searchReddit(queryBundle.hypothesis),
        // Skip patents.google when geography is set — global patents drown local signals
        geography
          ? searchWeb(`${primaryQuery} startup OR platform OR initiative`, locale)
          : searchWeb(`${product} patent filing technology site:patents.google.com OR site:patents.justia.com`, locale),
        ...entityProbes.slice(0, 1).map(q => searchWeb(q, locale)),
      ]),
    ingest: (settled, round) => {
      const sources: AgentSource[] = [];
      const rawContent: string[] = [];
      const allHits: SearchResult[] = [];

      const takeSearch = (
        result: PromiseSettledResult<unknown> | undefined,
        label: string,
        tool: AgentSource['tool'],
        limit: number,
      ) => {
        if (!result || result.status !== 'fulfilled') return;
        const value = result.value as ToolResult<SearchResult[]>;
        if (!Array.isArray(value?.data)) return;
        const asSearch = value.data.map(r => ({
          title: r.title,
          url: r.url,
          snippet: 'snippet' in r ? String((r as SearchResult).snippet ?? '') : '',
        }));
        const { kept } = filterRelevant(asSearch, requirements, { limit, minScore: geography ? 0.25 : 0.15 });
        kept.forEach(r => {
          allHits.push(r);
          sources.push({ url: r.url, title: r.title, timestamp: value.timestamp, tool });
          rawContent.push(`[${label}] ${r.title}: ${r.snippet}`);
        });
      };

      if (round === 1) {
        takeSearch(settled[0], 'PLATFORM THREAT', 'serpapi', 4);
        takeSearch(settled[1], 'ADJACENT MARKET', 'serpapi', 4);
        takeSearch(settled[2], 'DISRUPTOR', 'serpapi', 4);
        takeSearch(settled[3], 'FUNDING', 'serpapi', 4);
        if (settled[4]?.status === 'fulfilled') {
          const hn = settled[4].value as { hnResult: ToolResult<Array<{ url: string; title: string; created: string }>>; summary: string };
          hn.hnResult?.data?.slice(0, 3).forEach(p => {
            sources.push({ url: p.url, title: p.title, timestamp: p.created, tool: 'hn' });
          });
          if (hn.summary) rawContent.push(`[HN TECH SENTIMENT] ${hn.summary}`);
        }
        takeSearch(settled[5], 'REDDIT ADJACENT', 'reddit', 4);
        takeSearch(settled[6], geography ? 'LOCAL THREAT' : 'PATENT SIGNAL', 'serpapi', 3);
        for (let i = 7; i < settled.length; i++) takeSearch(settled[i], 'ENTITY ADJACENT', 'serpapi', 3);
        candidates = extractCandidates(allHits, { geographyName: geography?.name, exclude: [product], limit: 5 });
      } else {
        for (const s of settled) {
          if (s.status !== 'fulfilled') continue;
          const val = s.value as Awaited<ReturnType<typeof discoverAndScrape>> | ToolResult<SearchResult[]>;
          if (val && 'pages' in val && Array.isArray(val.pages)) {
            val.pages.forEach(pageResult => {
              if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
              sources.push({ url: pageResult.data.url, title: pageResult.data.title || 'Candidate', timestamp: pageResult.timestamp, tool: 'firecrawl' });
              rawContent.push(`[CANDIDATE] ${pageResult.data.title}: ${pageResult.data.excerpt}`);
            });
          } else if (val && 'data' in val && Array.isArray(val.data)) {
            takeSearch(
              { status: 'fulfilled', value: val } as PromiseFulfilledResult<ToolResult<SearchResult[]>>,
              'GAP QUERY',
              'serpapi',
              4,
            );
          }
        }
      }

      return {
        sources,
        rawContent,
        toolResults: extractToolResults(settled as PromiseSettledResult<ToolResult<unknown>>[]),
        searchedFor,
        relevantSourceCount: sources.length,
        relevantHits: allHits,
        scrapedPageCount: sources.filter(s => s.tool === 'firecrawl').length,
      };
    },
    gapRound: async () => {
      if (gapQueries.length > 0) {
        const settled = await Promise.allSettled(
          gapQueries.slice(0, 2).map(q => searchWeb(q, locale)),
        );
        searchedFor.push(...gapQueries.slice(0, 2));
        return settled;
      }

      const { settled, queries } = await verifyCandidates(candidates, {
        product,
        geographyName: geography?.name,
        category: cat ?? categoryLabel,
        maxCandidates: 2,
        topN: 1,
      });
      searchedFor.push(...queries);
      return settled;
    },
  });

  if (candidates.length) loop.evidence = { ...loop.evidence, candidates };

  const systemPrompt = `You are a strategic threat analyst identifying companies from OUTSIDE the primary category that could disrupt it.
${priorContext ? `\nPrior conversation context:\n${priorContext}` : ''}
${evidencePromptRules(geography, cat)}`;

  const userPrompt = `Query: "${query}"
Product category: ${categoryLabel}
${geography ? `Geography: ${geography.name}` : ''}
Evidence status: ${loop.evidence.status}

Raw signals:
${(await import('@/lib/guardrails')).fenceUntrusted(loop.rawContent)}

Produce JSON:
{
  "insufficientEvidence": boolean,
  "facts": string[],
  "interpretation": string[],
  "threats": [{ "company": string, "category": string, "threatVector": string, "riskLevel": "high" | "medium" | "low", "evidence": string }],
  "overallRisk": "high" | "medium" | "low",
  "timeToImpact": string,
  "defensiveActions": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}

If insufficientEvidence, leave threats empty — do not invent global patent threats for a local market.`;

  let parsed: Record<string, unknown> = {};
  try {
    parsed = await generateHuggingFaceJson(systemPrompt, userPrompt, { maxNewTokens: 1400, temperature: 0.2 });
  } catch {
    parsed = { insufficientEvidence: true, facts: [], interpretation: loop.evidence.gaps, threats: [], overallRisk: 'medium', timeToImpact: '12-18 months', defensiveActions: [], confidenceScore: 0.35 };
  }

  const gate = applyInsufficientGate(parsed, loop.evidence);
  if (gate.insufficient) loop.evidence = { ...loop.evidence, status: 'insufficient' };

  const confScore = Number.parseFloat(
    (gate.confidenceScore * computeSignalQualityPenalty(loop.toolResults, 7)).toFixed(2),
  );

  const output: AdjacentOutput = {
    agentId: 'adjacent',
    domain: 'adjacent',
    artifactType: 'threat-heatmap',
    confidence: scoreToLevel(confScore),
    confidenceScore: confScore,
    facts: gate.insufficient ? [] : (parsed.facts as string[] ?? []),
    interpretation: gate.insufficient ? loop.evidence.gaps : (parsed.interpretation as string[] ?? []),
    sources: loop.sources,
    generatedAt: new Date().toISOString(),
    threats: gate.insufficient ? [] : ((parsed.threats ?? []) as AdjacentThreat[]),
    overallRisk: (parsed.overallRisk as AdjacentOutput['overallRisk']) ?? 'medium',
    timeToImpact: (parsed.timeToImpact as string) ?? '12-18 months',
    defensiveActions: gate.insufficient ? [] : (parsed.defensiveActions as string[] ?? []),
    evidence: loop.evidence,
    toolCallCount: loop.toolCallCount,
    searchCallCount: loop.searchCallCount,
    scrapeCallCount: loop.scrapeCallCount,
    droppedIrrelevantCount: loop.droppedIrrelevantCount,
  };

  return output;
}

export const adjacentAgent: AgentConfig = {
  id: 'adjacent',
  name: 'Adjacent Threat Agent',
  description: 'Identifies companies from outside the category that could disrupt it — platform expansion, infrastructure players, category convergence.',
  run,
};

File: backend/app/agents/adjacent.py
"""Adjacent Market Collision agent (parity with lib/agents/adjacent.ts)."""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from typing import Any

from app.gemini import generate_json
from app.models import score_to_level
from app.tools.fallback import compute_signal_quality_penalty, extract_tool_results
from app.tools.hn_algolia import get_tech_sentiment
from app.tools.reddit import search_reddit
from app.tools.serpapi import search_news, search_web

AGENT_ID = "adjacent"
AGENT_NAME = "Adjacent Threat Agent"
AGENT_DESCRIPTION = (
    "Identifies companies from outside the category that could disrupt it — "
    "platform expansion, infrastructure players, category convergence."
)


def _fulfilled(result: Any) -> bool:
    return not isinstance(result, BaseException)


def _add_web_results(
    result: Any,
    label: str,
    sources: list[dict[str, Any]],
    raw_content: list[str],
) -> None:
    if not _fulfilled(result):
        return
    for r in (result.get("data") or [])[:4]:
        sources.append(
            {
                "url": r["url"],
                "title": r["title"],
                "timestamp": result["timestamp"],
                "tool": "serpapi",
            }
        )
        raw_content.append(f"[{label}] {r['title']}: {r['snippet']}")


async def run(ctx: dict[str, Any]) -> dict[str, Any]:
    query = ctx["query"]
    product = ctx["product"]
    competitor = ctx.get("competitor")
    prior_context = ctx.get("prior_context")

    category = f"{product} vs {competitor}" if competitor else product

    (
        platform_threat_result,
        ai_funding_result,
        disruptor_result,
        funding_result,
        hn_adjacent_result,
        reddit_adjacent_result,
    ) = await asyncio.gather(
        search_web(f"{product} competitors alternatives disruption market 2025 2026"),
        search_web(f"{product} category adjacent market expansion threat 2025"),
        search_web(f"companies replacing {product} OR disrupting {category} 2025 2026"),
        search_news(f"{product}{f' {competitor}' if competitor else ''} market disruption funding threat"),
        get_tech_sentiment(f"{product} disruption threat"),
        search_reddit(f"{product} alternatives what are people using instead"),
        return_exceptions=True,
    )

    (patent_result,) = await asyncio.gather(
        search_web(
            f"{product} patent filing technology site:patents.google.com OR site:patents.justia.com"
        ),
        return_exceptions=True,
    )

    sources: list[dict[str, Any]] = []
    raw_content: list[str] = []

    _add_web_results(platform_threat_result, "PLATFORM THREAT", sources, raw_content)
    _add_web_results(ai_funding_result, "ADJACENT MARKET", sources, raw_content)
    _add_web_results(disruptor_result, "DISRUPTOR", sources, raw_content)
    _add_web_results(funding_result, "FUNDING", sources, raw_content)

    if _fulfilled(hn_adjacent_result):
        hn_data = hn_adjacent_result.get("hnResult") or hn_adjacent_result.get("hn_result") or {}
        summary = hn_adjacent_result.get("summary", "")
        for p in (hn_data.get("data") or [])[:3]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "hn",
                }
            )
        raw_content.append(f"[HN TECH SENTIMENT] {summary}")

    if _fulfilled(reddit_adjacent_result):
        for p in (reddit_adjacent_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "reddit",
                }
            )
            raw_content.append(f"[REDDIT ADJACENT] {p['title']}: {p['snippet']}")

    _add_web_results(patent_result, "PATENT SIGNAL", sources, raw_content)

    prior_block = f"\nPrior conversation context:\n{prior_context}" if prior_context else ""
    system_prompt = f"""You are a strategic threat analyst who identifies companies from OUTSIDE the primary category that could disrupt it. You think in terms of market adjacency, platform expansion, and category convergence.

Key question: What companies or trends are NOT currently in the {category} space but have the distribution, data, or technology to enter it or displace it credibly within 12-18 months?

Types of adjacent threats to watch:
1. Platform expansion — large platforms adding the same capability as a feature
2. Infrastructure players — lower-level tech companies moving up-stack
3. Horizontal AI — general-purpose AI agents expanding into this vertical
4. Category convergence — adjacent tools expanding into this space{prior_block}"""

    user_prompt = f"""Query: "{query}"
Product category: {category}

Raw signals:
{chr(10).join(raw_content)}

Produce JSON:
{{
  "facts": string[],
  "interpretation": string[],
  "threats": [
    {{
      "company": string,
      "category": string,
      "threatVector": string,
      "riskLevel": "high" | "medium" | "low",
      "evidence": string
    }}
  ],
  "overallRisk": "high" | "medium" | "low",
  "timeToImpact": string,
  "defensiveActions": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}}"""

    try:
        parsed = await generate_json(system_prompt, user_prompt, max_new_tokens=1400, temperature=0.2)
    except Exception:
        parsed = {
            "facts": [
                re.sub(r"^\[[^\]]+\]\s*", "", s)
                for s in raw_content[:3]
                if len(re.sub(r"^\[[^\]]+\]\s*", "", s)) > 15
            ],
            "interpretation": [
                "Analysis synthesis is temporarily unavailable. Raw data signals are shown below."
            ],
            "threats": [],
            "overallRisk": "medium",
            "timeToImpact": "12-18 months",
            "defensiveActions": [],
            "synthesizedAnswer": "Adjacent threat data collected but synthesis failed.",
            "confidenceScore": 0.4,
        }

    raw_score = parsed["confidenceScore"] if isinstance(parsed.get("confidenceScore"), (int, float)) else 0.6
    tool_results = extract_tool_results(
        [
            platform_threat_result,
            ai_funding_result,
            disruptor_result,
            funding_result,
            hn_adjacent_result,
            reddit_adjacent_result,
            patent_result,
        ]
    )
    conf_score = round(float(raw_score) * compute_signal_quality_penalty(tool_results, 7), 2)
    confidence = score_to_level(conf_score)

    return {
        "agentId": AGENT_ID,
        "domain": "adjacent",
        "artifactType": "threat-heatmap",
        "confidence": confidence,
        "confidenceScore": conf_score,
        "facts": parsed.get("facts") or [],
        "interpretation": parsed.get("interpretation") or [],
        "sources": sources,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "threats": parsed.get("threats") or [],
        "overallRisk": parsed.get("overallRisk") or "medium",
        "timeToImpact": parsed.get("timeToImpact") or "12-18 months",
        "defensiveActions": parsed.get("defensiveActions") or [],
    }
